### PR TITLE
Make tasks more flexible

### PR DESCRIPTION
**Pull request by ejnnr**
Mainly, we'd like to allow detectors that require both trusted and untrusted data (where untrusted data is an unlabeled mix of clean and anomalous data). I ended up making a bunch of other small changes that seemed good while I was doing this:
- Removed `TaskConfigBase`, it was pretty pointless
- Added a `get_test_split()` method to `DatasetConfig`---this will sometimes raise NotImplementedError but for existing datasets we can always make it work, and it's useful when defining tasks. I don't really like the design though, maybe we should just explicitly have "dataset pair" objects?
- Modified adversarial examples a bit to provide test splits
- Replaced the `RemoveLabelMixDataset` with an argument to `TestDataMix` (now `MixedData` since it's not just used for tests). Probably simpler this way

Demo notebook and tests pass but I didn't add new ones to actually test all the mixing logic etc. is correct, just did some manual sanity checks.

@VRehnberg Let me know if you have objections to the design or better ideas for what it should look like, otherwise we can merge this.

### CODE REVIEW

**Comment by ejnnr** [comment]
Hm, I should definitely let the trusted and untrusted clean training data be different before merging, otherwise it's way too easy.

File: src/cupbearer/tasks/_config.py
Comment (VRehnberg) [inline src/cupbearer/tasks/_config.py:131]: I suppose splitting this way is more realistic in a sense.

**Review by VRehnberg — APPROVED** [review]
LGTM, though haven't run the test-suite yet but feel to merge if you've made sure the test-suite runs for you.

**Comment by VRehnberg** [comment]
I've found some things so hold off on merging.

File: src/cupbearer/tasks/adversarial_examples.py
Comment (VRehnberg) [inline src/cupbearer/tasks/adversarial_examples.py:25]:
```suggestion
    def _get_anomalous_data(self) -> DatasetConfig:
        return self._get_anomalous_data_by_split(train=True)

    def _get_anomalous_test_data(self) -> DatasetConfig:
        return self._get_anomalous_data_by_split(train=False)

    def _get_anomalous_data_by_split(self, train: bool) -> DatasetConfig:
```
Something like this to match signature in parent class.

File: tests/test_pipeline.py
Comment (VRehnberg) [inline tests/test_pipeline.py]:
```suggestion
```
Backdoor is no longer an argument to tasks. This is the same for all the below tests as well.

File: src/cupbearer/tasks/backdoor_detection.py
Comment (VRehnberg) [inline src/cupbearer/tasks/backdoor_detection.py:20]: The name change from `backdoor` to `_backdoor` breaks the `test_wanet` test. The test should be updated.

**Review by VRehnberg — CHANGES_REQUESTED** [review]
Found some things as I was running the test suite. Might be things beyond these as well, but as #30 might supercede this PR I left it at these.

**Comment by ejnnr** [comment]
Closing this in favor of #30 (I think the changes there are good overall and even if we iterate on them a bit, we won't go back to this version)

**Pull request closed without merging**
